README
This package is http://pear.php.net/package/Net_FTP and has been migrated from http://svn.php.net/repository/pear/packages/Net_FTP

Please report all new issues via the PEAR bug tracker.

If this package is marked as unmaintained and you have fixes, please submit your pull requests and start discussion on the pear-qa mailing list.

To test, run either
$ phpunit tests/
  or
$ pear run-tests -r

To build, simply
$ pear package

To install from scratch
$ pear install package.xml

To upgrade
$ pear upgrade -f package.xml


CHANGELOG
1.4.0a3 - 2008-??-??
* Fixed Bug #13946: Test Failed : testPutRecursive
* Fixed Bug #14513: Timestamp when using the LS function
* Implemented Request #14773: setPassive possible before connecting

1.4.0a2 - 2008-05-19
* Fixed Bug #13496: set bit not supported
* Fixed Bug #13689: . in file owner or group name breaks _ls_match
* Fixed Bug #13690: getRecursive does not work because of '.' and '..' directories
* Fixed Bug #13114: Error check in $FTP->getRecursive(...)
* Implemented Request #13104: connect() <> $this->handle in login()
* Fixed Bug #13917: putRecursive always return PEAR::Error
* Fixed Bug #13850: inclusion of Net/FTP.php in tests doesn't always work
* Implemented Request #2271: New flag for _rm_dir_recursive to leave directory structure intact
* Implemented Request #6806: Allow recursive up-/download with only adding newer files

1.4.0a1 - 2007-12-22
* Fixed bug #6754: Add method to set directory list matcher explicitly
* Fixed Bug #7102: Add ability to disable use of ftp_nb_put(), can sometimes be very slow

* Improved detection of remote directories. It is slower than before, but does not
rely on simply detecting if the given path ends with a slash.

1.3.7 - 2008-05-19
* Fixed Bug #13917: putRecursive always return PEAR::Error
* Fixed Bug #13850: inclusion of Net/FTP.php in tests doesn't always work

1.3.6 - 2008-05-06
* Fixed Bug #13496: set bit not supported
* Fixed Bug #13689: . in file owner or group name breaks _ls_match
* Fixed Bug #13690: getRecursive does not work because of '.' and '..' directories
* Fixed Bug #13114: Error check in $FTP->getRecursive(...)

1.3.5 - 2008-01-13
* Fixed Bug #12840: Fix compliance with PHP 4 (broken since 1.3.3)

1.3.4 - 2007-12-10
* Fixed Bug #12639: _constructPath() prevents _checkDir() from working correctly

1.3.3 - 2007-12-06
* Fixed Bug #7146: Recursive mkdir() broken on Windows
* Fixed Bug #7270: Recursive rmdir() broken
* Fixed Bug #7527: ls fails if there are no files and a total line
* Fixed Bug #8102: Loading file extension and checking extension gives binary for ascii files
* Fixed Bug #9611: (, ? and ) break detection of the unix platform
* Fixed Bug #10237: put() doesn't run ftp_alloc to allocate space
* PEAR Coding Style Valid
* Removed package.xml version 1.0

1.3.2 - 2006-02-11
* Fixed Bug #4102: Problem detecting os method ls().
* Fixed Bug #5337: _list_and_parse behavior with an empty remote directory.
* Fixed Bug #4836: Off-by-one error in regex for Windows directory listings.
* Fixed Bug #4749: ls() fails when connection is closed.
* Fixed Bug #4969: Recursive rm ends in endless loop.
* Fixed Bug #5895: Recursive chmod ends in endless loop

1.3.1 - 2005-04-07 
* Fix bug #3778 "Notice of Uninitialized string offset in function _rm_dir_recursive"
* Fix bug #4008 "_list_and_parse tries to determine OS on an empty list"
* Fix bug #4009 "_determine_os_match doesn't take into account numbered users and groups"

1.3.0 - 2005-03-31
- Updated year.
- Updated docblocks regarding the new standard.
- Fixed whitespace issues.
- Fixed bug 3362: bug in Net_FTP::_rm_file.
- Added PEAR 1.4 compatible package2.xml (package.xml version 2.0).

1.3.0RC2 2005-01-04
* Added example for observer.
* Fixed CS issues.
* Fixed bug #2654: mkdir() should not fail if the directory already exists.
* Fixed bug #2648: if ls() returns array(), net_FTP bails.
* Fixed some unknown small bugs.

1.3.0RC1 2004-09-23
* Fixed bug #2268: getRecursive does not work because of '.' and '..' directories. 

1.3.0beta4 2004-06-28
* Optimized OS determination. Thanks to Eric Quilantang. 
* Added ability to set timeout in constructor.
* Added much more error handling.
* Added real error codes with full documentation.
* Fix bug #1176: Net_FTP::ls can return dates in the future. 
* Fix bug #1446: Bug in isPassive().
* Fix bug #1480 _list_and_parse() may throw warnings.
* Fix bug #1445 Improvement of setTimeout() method + timout handling.
* Renamed makeDirPermissions() to _makeDirPermissions(), since it's private (CS).
* Fix './' syntax for pathes.

1.3.0beta3 2004-05-25
* Added "t" for permission regex (some servers mark speacial dirs).
* Added better support for different ls formats (thanks for hint by George Kearns).

1.3.0beta2 2004-03-20
* Added patch by Ilja Polivanovas <ipa@assis.lt> to enable correct directory permissions.
* Added rename() functionality. Thanks to the unnamed coder from "Webmestre ESGI" <wmaster_esgi@hotmail.com>.

1.3.0beta1 2004-03-14
* Fixed bug #336: ftp_login displays Warning.
* Fixed bug #813: ftp_nb_put function abort when mode is null.
* Fixed bug #240: mkdir() with $recursive===true doesn't support "single" directories.
* Fixed bug #723: Timeout for connect cannot be set.
* Fixed bug #918: getRecursive() corrupts binary file.
* Fixed typos.

1.2 2003-09-29
Fixes some typos and bugs.

1.1 2003-07-22
Fix for FatalError on Line 834 (rmrecursive()).

1.0 2003-07-15
- Fixed different typos in code, comments and example.
- Fixed recursive deletion trying to remove "." and ".." entries in the directorylist.
- Fixed uploading above 4096 byte, when using ftp_nb_put()

0.9 2003-02-22
Unknown

0.5 2002-12-20
First release


tests/AllTests.php
<?php
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * Net_FTP test suite file
 *
 * To run the tests execute this from the top directory for a CVS checkout
 * $ pear run-tests -ur
 * or this if you've installed the package
 * $ pear run-tests -pu Net_FTP
 *
 * PHP version 5
 *
 * LICENSE: This source file is subject to version 3.0 of the PHP license
 * that is available through the world-wide-web at the following URI:
 * http://www.php.net/license/3_0.txt.  If you did not receive a copy of
 * the PHP License and are unable to obtain it through the web, please
 * send a note to license@php.net so we can mail you a copy immediately.
 *
 * @category  Networking
 * @package   FTP
 * @author    Tobias Schlitt <toby@php.net>
 * @copyright 1997-2008 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   CVS: $Id$
 * @link      http://pear.php.net/package/Net_FTP
 * @link      http://www.phpunit.de PHPUnit
 * @since     File available since Release 0.0.1
 */

if (!defined('PHPUnit_MAIN_METHOD')) {
    define('PHPUnit_MAIN_METHOD', 'Net_FTP_AllTests::main');
}

require_once 'PHPUnit/Framework/TestSuite.php';
require_once 'PHPUnit/TextUI/TestRunner.php';

chdir(dirname(__FILE__));
require_once 'Net_FTPTest.php';

/**
 * Unit test case for Net_FTP
 *
 * @category  Networking
 * @package   FTP
 * @author    Jorrit Schippers <jschippers@php.net>
 * @copyright 1997-2008 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   Release: @package_version@
 * @link      http://pear.php.net/package/Net_FTP
 * @since     Class available since Release 1.3.3
 */
class Net_FTP_AllTests
{
    /**
     * Main test suite run method
     *
     * @return void
     */
    public static function main()
    {
        PHPUnit_TextUI_TestRunner::run(self::suite());
    }

    /**
     * Main test suite method
     *
     * @return void
     */
    public static function suite()
    {
        $suite = new PHPUnit_Framework_TestSuite('Net_FTP Tests');
        $suite->addTestSuite('Net_FTPTest');

        return $suite;
    }
}

if (PHPUnit_MAIN_METHOD == 'Net_FTP_AllTests::main') {
    Net_FTP_AllTests::main();
}
?>

tests/Net_FTPTest.php
<?php
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * Net_FTP main test file
 *
 * To run the tests either execute this from the top directory if checked out from
 * CVS:
 * $ pear run-tests -ur 
 * or if you want to run the tests on an installed version, run from within any
 * directory:
 * $ pear run-tests -pu Net_FTP
 *
 * In both cases you need PHPUnit installed
 *
 * PHP version 5
 *
 * LICENSE: This source file is subject to version 3.0 of the PHP license
 * that is available through the world-wide-web at the following URI:
 * http://www.php.net/license/3_0.txt.  If you did not receive a copy of
 * the PHP License and are unable to obtain it through the web, please
 * send a note to license@php.net so we can mail you a copy immediately.
 *
 * @category  Networking
 * @package   FTP
 * @author    Tobias Schlitt <toby@php.net>
 * @copyright 1997-2008 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   CVS: $Id$
 * @link      http://pear.php.net/package/Net_FTP
 * @link      http://www.phpunit.de PHPUnit
 * @since     File available since Release 1.3.3
 */

if (!defined('PHPUnit_MAIN_METHOD')) {
    define('PHPUnit_MAIN_METHOD', 'Net_FTPTest::main');
}

require_once 'PHPUnit/Framework.php';
require_once 'System.php';

chdir(dirname(__FILE__));
if (substr(dirname(__FILE__), -6) == DIRECTORY_SEPARATOR.'tests') {
    include_once '../Net/FTP.php';
} else {
    include_once 'Net/FTP.php';
}

/**
 * Unit test case for Net_FTP
 *
 * @category  Networking
 * @package   FTP
 * @author    Jorrit Schippers <jschippers@php.net>
 * @copyright 1997-2008 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   Release: @package_version@
 * @link      http://pear.php.net/package/Net_FTP
 * @since     Class available since Release 1.3.3
 */
class Net_FTPTest extends PHPUnit_Framework_TestCase
{
    protected $ftp;
    protected $ftpdir;
    protected $setupError;
    
    /**
     * Runs the test methods of this class.
     *
     * @access public
     * @static
     * @return void
     */
    public static function main()
    {
        include_once 'PHPUnit/TextUI/TestRunner.php';

        $suite  = new PHPUnit_Framework_TestSuite('Net_FTPTest');
        PHPUnit_TextUI_TestRunner::run($suite);
    }

    /**
     * Sets up the fixture, for example, opens a network connection.
     * This method is called before a test is executed.
     *
     * @access protected
     * @return void
     */
    protected function setUp()
    {
        if (!file_exists('config.php')) {
            $this->setupError = 'config.php does not exist in '.getcwd();
            return;
        }
        
        include_once 'config.php';
        
        if (!defined('FTPHOST') || !defined('FTPPORT') || !defined('FTPUSER')
            || !defined('FTPPASSWORD')) {
            $this->setupError = 'Some required constants are not defined';
            return;
        }
        
        $this->ftp = new Net_FTP(FTPHOST, FTPPORT, 30);
        
        $res = $this->ftp->login(FTPUSER, FTPPASSWORD);
        
        if (PEAR::isError($res)) {
            $this->setupError = 'Could not login to the FTP server';
            $this->ftp        = null;
            return;
        }
        
        if (defined('FTPDIR') && '' !== FTPDIR) {
            $res = $this->ftp->cd(FTPDIR);
            
            if (PEAR::isError($res)) {
                $this->setupError = 'Could switch to directory '.FTPDIR;
                $this->ftp        = null;
                return;
            }
        }
        
        $res = $this->ftp->pwd();
        
        if (PEAR::isError($res)) {
            $this->setupError = 'Could not get current directory';
            $this->ftp        = null;
            return;
        }
        
        $this->ftpdir = $res;
        
        $res = $this->ftp->mkdir('test');
        
        if (PEAR::isError($res)) {
            $this->setupError = 'Could not create a test directory';
            $this->ftp        = null;
            return;
        }
        
        $res = $this->ftp->cd('test');
        
        if (PEAR::isError($res)) {
            $this->setupError = 'Could not change to the test directory';
            $this->ftp        = null;
            return;
        }
    }

    /**
     * Tears down the fixture, for example, closes a network connection.
     * This method is called after a test is executed.
     *
     * @access protected
     * @return void
     */
    protected function tearDown()
    {
        if ($this->ftp != null) {
            $this->ftp->cd($this->ftpdir);
            $this->ftp->rm('test/', true);
            $this->ftp->disconnect();
            
            $this->ftpdir     = null;
            $this->ftp        = null;
            $this->setupError = null;
        }
    }
    
    /**
     * Tests functionality of Net_FTP::mkdir()
     * 
     * @return void
     * @see Net_FTP::mkdir()
     */
    public function testMkdir()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $this->ftp->mkdir('dir1', false);
        $this->ftp->mkdir('dir1/dir2/dir3/dir4', true);
        $this->assertTrue($this->ftp->cd('dir1/dir2/dir3/dir4'));
    }
    
    /**
     * Tests functionality of Net_FTP::mkdir()
     * 
     * @return void
     * @see Net_FTP::mkdir()
     */
    public function testRename()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $this->ftp->put('testfile.dat', 'testfile.dat', FTP_ASCII);
        $this->assertTrue($this->ftp->rename('testfile.dat', 'testfile2.dat'));
    }

    /**
     * Tests functionality of Net_FTP::rm()
     * 
     * @return void
     * @see Net_FTP::rm()
     */
    public function testRm()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $list1 = $this->ftp->ls();
        
        $this->ftp->put('testfile.dat', 'testfile.dat', FTP_ASCII);
        $this->ftp->mkdir('dir1/dir2/dir3/dir4', true);
        
        $this->ftp->rm('dir1/', true);
        $this->ftp->rm('testfile.dat');
        
        $list2 = $this->ftp->ls();
        
        $this->assertEquals($list1, $list2, 'Directory listing before creation and'.
            ' after creation are not equal');
        
        $this->ftp->mkdir('dir1');
        $this->ftp->cd('dir1');
        $this->ftp->mkdir('dir1');
        $this->ftp->mkdir('dir2');
        $this->ftp->mkdir('dir3');
        $this->ftp->put('testfile.dat', 'testfile.dat', FTP_ASCII);
        
        $this->ftp->cd('..');
        
        $this->ftp->rm('dir1/', true, true);
        
        $this->ftp->cd('dir1');
        
        $list = $this->_getNames($this->ftp->ls());
        $dirs = array('dir1', 'dir2', 'dir3');
        
        sort($list);
        sort($dirs);
        
        $this->assertEquals($list, $dirs, 'The resulting directory listing is not '.
            'correct');
    }

    /**
     * Tests functionality of Net_FTP::putRecursive()
     *
     * @return void
     * @see Net_FTP::putRecursive()
     */
    public function testPutRecursive()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        
        $tmpdir    = array();
        $tmpfile   = array();
        $tmpdir[]  = System::mktemp(array('-d', 'pearnetftptest'));
        $tmpdir[]  = System::mktemp(array('-t', $tmpdir[0], '-d'));
        $tmpdir[]  = System::mktemp(array('-t', $tmpdir[1], '-d'));
        $tmpfile[] = System::mktemp(array('-t', $tmpdir[0]));
        $tmpfile[] = System::mktemp(array('-t', $tmpdir[1]));
        $tmpfile[] = System::mktemp(array('-t', $tmpdir[2]));
        
        $local  = $tmpdir[0].DIRECTORY_SEPARATOR;
        $remote = './'.$this->_getLastPart($tmpdir[0]).'/';
        
        $ret = $this->ftp->putRecursive($local, $remote);
        $this->assertFalse(PEAR::isError($ret));
        
        for ($i = 0; $i < 3; $i++) {
            $ret = $this->ftp->cd($this->_getLastPart($tmpdir[$i]).'/');
            $this->assertFalse(PEAR::isError($ret));
            
            $dirlist = $this->ftp->ls();
            $this->assertFalse(PEAR::isError($dirlist));
          
            $dirlist   = $this->_getNames($dirlist);
            $dirlistok = array($this->_getLastPart($tmpfile[$i]));
            if ($i < 2) {
                $dirlistok[] = $this->_getLastPart($tmpdir[$i+1]);
            }
            
            sort($dirlist);
            sort($dirlistok);
            
            $this->assertEquals($dirlist, $dirlistok);
        }
    }

    /**
     * Tests functionality of Net_FTP::put()
     * 
     * @return void
     * @see Net_FTP::put()
     */
    public function testPut()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $res = $this->ftp->put('testfile.dat', 'testfile.dat', true, FTP_ASCII,
            NET_FTP_BLOCKING | NET_FTP_NONBLOCKING);
        $this->assertTrue(PEAR::isError($res), 'No error return while it should');
        if (PEAR::isError($res)) {
            $this->assertEquals($res->getCode(), NET_FTP_ERR_BADOPTIONS,
                'Bad error code returned');
        }
    }

    /**
     * Tests functionality of Net_FTP::_makeDirPermissions()
     * 
     * @return void
     * @see Net_FTP::_makeDirPermissions()
     */
    public function testMakeDirPermissions()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $tests = array(
            '111' => '111',
            '110' => '110',
            '444' => '555',
            '412' => '512',
            '641' => '751',
            '666' => '777',
            '400' => '500',
            '040' => '050',
            '004' => '005',
        );
        
        foreach ($tests AS $in => $out) {
            $this->assertEquals($this->ftp->_makeDirPermissions($in), $out);
        }
    }

    /**
     * Tests functionality of Net_FTP::size()
     * 
     * @return void
     * @see Net_FTP::size()
     */
    public function testSize()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        // upload in binary to avoid addition/removal of characters
        $this->ftp->put('testfile.dat', 'testfile.dat', FTP_BINARY);
        $this->assertEquals($this->ftp->size('testfile.dat'),
            filesize('testfile.dat'));
    }
    
    /**
     * Tests functionality of Net_FTP::setMode(), Net_FTP::checkFileExtension(),
     * Net_FTP::addExtension() and Net_FTP::removeExtension()
     * 
     * @return void
     * @see Net_FTP::checkFileExtension(), Net_FTP::addExtension(),
     *      Net_FTP::removeExtension(), Net_FTP::setMode()
     */
    public function testExtensions()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $this->ftp->setMode(FTP_ASCII);
        $this->ftp->addExtension(FTP_BINARY, 'tst');
        $this->assertEquals($this->ftp->checkFileExtension('test.tst'), FTP_BINARY);
        $this->ftp->removeExtension('tst');
        $this->assertEquals($this->ftp->checkFileExtension('test.tst'), FTP_ASCII);
        $this->ftp->setMode(FTP_BINARY);
        $this->assertEquals($this->ftp->checkFileExtension('test.tst'), FTP_BINARY);
    }

    /**
     * Tests functionality of Net_FTP::getExtensionsFile()
     * 
     * @return void
     * @see Net_FTP::getExtensionsFile()
     */
    public function testGetExtensionsFile()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $res = $this->ftp->getExtensionsFile('extensions.ini');
        $this->assertFalse(PEAR::isError($res), 'Test extensions file could be'.
            'loaded');
        
        $this->ftp->setMode(FTP_BINARY);
        $this->assertEquals($this->ftp->checkFileExtension('test.asc'), FTP_ASCII);
        $this->ftp->setMode(FTP_ASCII);
        $this->assertEquals($this->ftp->checkFileExtension('test.gif'), FTP_BINARY);
    }

    /**
     * Tests functionality of Net_FTP::setDirMatcher()
     * 
     * @since 1.4.0a1
     * @return void
     */
    public function testSetDirMatcher()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        
        $pattern = '/(?:(d)|.)([rwxt-]+)\s+(\w+)\s+(\w+)\s+(\w+)\s(\w+\s+\d+\s+.+)'.
            '\s+(.+)/';
        $map     = array(
            'is_dir' => 1,
            'rights' => 2,
            'user' => 3,
            'group' => 4,
            'size' => 5,
            'date' => 6,
            'name' => 7,
        );
        
        $res = $this->ftp->setDirMatcher(false, false);
        $this->assertTrue(PEAR::isError($res)
            && $res->getCode() == NET_FTP_ERR_ILLEGALPATTERN,
            'Result of setDirMatcher(boolean, boolean) should give a '.
            'NET_FTP_ERR_ILLEGALPATTERN error');
        
        $res = $this->ftp->setDirMatcher($pattern, false);
        $this->assertTrue(PEAR::isError($res)
            && $res->getCode() == NET_FTP_ERR_ILLEGALMAP,
            'Result of setDirMatcher(string, boolean) should give a '.
            'NET_FTP_ERR_ILLEGALMAP error');
        
        $res = $this->ftp->setDirMatcher($pattern, array('a' => 'b'));
        $this->assertTrue(PEAR::isError($res)
            && $res->getCode() == NET_FTP_ERR_ILLEGALMAPVALUE,
            'The items in a matcher map should only contain numeric values');
        
        $res = $this->ftp->setDirMatcher($pattern, $map);
        $this->assertFalse(PEAR::isError($res),
            'A valid pattern and map should return no error');
    }
    
    /**
     * Tests changes made to fix bug #9611
     * 
     * @since 1.4.0a1
     * @return void
     */
    public function testCheckRemoteDir()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        
        $res = $this->ftp->_checkRemoteDir('test');
        $this->assertFalse($res, 'test should not be a directory');
        
        $this->ftp->put('testfile.dat', 'test', FTP_BINARY);
        $res = $this->ftp->_checkRemoteDir('test');
        $this->assertFalse($res, 'test should not be a directory');
        
        $this->ftp->rm('test');
        
        $this->ftp->mkdir('test');
        $res = $this->ftp->_checkRemoteDir('test');
        $this->assertTrue($res, 'test should be a directory');
    }

    /**
     * Tests _determineOSMatch
     * 
     * @return void
     * @see Net_FTP::_determineOSMatch()
     */
    public function testDetermineOSMatch()
    {
        if ($this->ftp == null) {
            $this->fail('This test requires a working FTP connection. Setup '.
            'config.php with proper configuration parameters. ('.
            $this->setupError.')');
        }
        $dirlist = array(
            'drwxrwsr-x  75 upload.  (?).         3008 Oct 30 21:09 ftp1',
        );
        
        $res = $this->ftp->_determineOSMatch($dirlist);
        $this->assertFalse(PEAR::isError($res),
            'The directory listing should be recognized');
        
        $this->assertEquals($res['pattern'],
            $this->ftp->_ls_match['unix']['pattern'],
            'The input should be parsed by the unix pattern');
    }
    
    /**
     * Return all name keys in the elements of an array, leaving out . and ..
     *
     * @param array $in Multidimensional array
     *
     * @return array Array containing name keys
     */
    function _getNames($in)
    {
        $return = array();
        foreach ($in as $v) {
            if ($v['name'] != '.' && $v['name'] != '..') {
                $return[] = $v['name'];
            }
        }
        return $return;
    }
     
    /**
     * Return the last element of a local path
     *
     * @param string $in Path
     *
     * @return array Last part of path
     */
    function _getLastPart($in)
    {
        $start = strrpos($in, DIRECTORY_SEPARATOR) + 1;
        return substr($in, $start);
    }
}

// Call Net_FTPTest::main() if this source file is executed directly.
if (PHPUnit_MAIN_METHOD == 'Net_FTPTest::main') {
    Net_FTPTest::main();
}
?>


example/socket.php
<pre>
<?php
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * Net_FTP_Socket example.
 *
 * Example for the usage of Net_FTP's socket implementation of the
 * ext/FTP functions.'
 *
 * PHP version 5
 *
 * LICENSE: This source file is subject to version 3.0 of the PHP license
 * that is available through the world-wide-web at the following URI:
 * http://www.php.net/license/3_0.txt.  If you did not receive a copy of
 * the PHP License and are unable to obtain it through the web, please
 * send a note to license@php.net so we can mail you a copy immediately.
 *
 * @category  Networking
 * @package   FTP
 * @author    Tobias Schlitt <toby@php.net>
 * @copyright 1997-2005 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   CVS: $Id$
 * @link      http://pear.php.net/package/Net_FTP
 * @since     File available since Release 1.3.0
 */

if (isset($_GET['native'])) {
    @dl('ftp.so');
}

// Configuration

// Login info
$host = 'localhost';
$user = 'pub';
$pass = 'public';

// Passive mode on/off
$pasv = false;

// List directory
$dir = 'episodes/';

// Upload files
$Uasci   = basename($_SERVER['PHP_SELF']);
$Ubinary = 'screenshot.jpg';

// End of configuration

$time = microtime(true);

/**
 * Function used by the test suit.
 * Spits out test results in readable way
 *
 * @param mixed $action action
 * @param mixed $result result
 * @param mixed $msg    message
 *
 * @return false
 */
function dump($action, $result, $msg = false)
{
    if (is_bool($result)) {
        $result = $result ? 'TRUE' : 'FALSE';
    }

    if (is_array($result)) {
        echo '<strong>' .$action. ':</strong>' ."\n";
        foreach ($result as $key => $value) {
            echo ' ' .$key. ': ' .$value. "\n";
        }
    } else {
        echo '<strong>' .$action. '</strong>:' ."\n";
        echo $result;
        if ($msg) {
            echo ' ( ' .$msg. ' )';
        }
        echo "\n";
    }

    echo '<hr style="border: 1px solid #000;"/>';// . "\n";
    flush();
}

/**
* Little test suit
*/
$stream = ftp_connect($host);
if (is_resource($stream)) {
    dump('Logging in', $bool = ftp_login($stream, $user, $pass));
    if ($bool) {
        dump('PWD', ftp_pwd($stream));
        dump('Systype', ftp_systype($stream));
        dump('CHDIR "'.$dir.'"', ftp_chdir($stream, $dir));
        dump('PWD', ftp_pwd($stream));
        dump('CDUP', ftp_cdup($stream));
        dump('PASSIVE', ftp_pasv($stream, $pasv));
        dump('RAWLIST "."', ftp_rawlist($stream, '.'));
        dump('CHMOD', ftp_chmod($stream, 0777, 'sfv3.php'));
        dump('ALLOCATE', ftp_alloc($stream, filesize($Ubinary), $msg), $msg);
        dump('UPLOAD ASCII', ftp_put($stream, $Uasci, $Uasci, FTP_ASCII), $Uasci);
        dump('UPLOAD BINARY', ftp_put($stream, $Ubinary, $Ubinary, FTP_BINARY),
            $Ubinary);
        dump('RAWLIST "."', ftp_rawlist($stream, '.'));
        dump('DELETE '.$Uasci, ftp_delete($stream, $Uasci));
        dump('DELETE '.$Bbinary, ftp_delete($stream, $Ubinary));
        dump('RAWLIST "."', ftp_rawlist($stream, '.'));
    }
    dump('QUIT', ftp_quit($stream));
}

$end = microtime(true);
echo $end-$time;
?>

</pre>


generate_package_xml.php
<?php
/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * File to generate a package.xml for a new Net_FTP release
 *
 * PHP versions 4 and 5
 *
 * LICENSE: This source file is subject to version 3.0 of the PHP license
 * that is available through the world-wide-web at the following URI:
 * http://www.php.net/license/3_0.txt.  If you did not receive a copy of
 * the PHP License and are unable to obtain it through the web, please
 * send a note to license@php.net so we can mail you a copy immediately.
 *
 * @category  Networking
 * @package   FTP
 * @author    Tobias Schlitt <toby@php.net>
 * @copyright 1997-2005 The PHP Group
 * @license   http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version   CVS: $Id$
 * @link      http://pear.php.net/package/Net_FTP
 * @since     File available since Release 1.3.0
 */

require_once 'PEAR/PackageFileManager2.php';

/**
 * Dump an error
 *
 * @param PEAR_Error $err The error object
 *
 * @return void
 */
function dumpError($err)
{
    echo $err->getMessage();
    exit;
}

$cvsdir     = '/cvs/pear/';
$packagedir = $cvsdir . 'Net_FTP/';

$current_version   = '1.4.0a3';
$current_stability = 'alpha';

$summary =
'Net_FTP provides an OO interface to the PHP FTP functions plus some additions';

$description =
'Net_FTP allows you to communicate with FTP servers in a more comfortable way
than the native FTP functions of PHP do. The class implements everything natively
supported by PHP and additionally features like recursive up- and downloading,
dircreation and chmodding. It also implements an observer pattern to allow
for example the view of a progress bar.';

$current_notes =
'* Fixed Bug #13946: Test Failed : testPutRecursive
* Fixed Bug #14513: Timestamp when using the LS function
* Implemented Request #14773: setPassive possible before connecting';

PEAR::setErrorHandling(PEAR_ERROR_CALLBACK, 'dumpError');

$p2 = new PEAR_PackageFileManager2();
$p2->setOptions(array(
    'baseinstalldir'    => '/',
    'filelistgenerator' => 'cvs',
    'packagedirectory'  => dirname(__FILE__),
    'include'           => array(),
    'ignore'            => array(
        'package.xml',
        'package2.xml',
        '*.tgz',
        'generate*',
        'doc*',
    ),
    'dir_roles'         => array(
        'tests'     => 'test',
        'example'   => 'doc'
    ),
    'simpleoutput'      => true,
));

$p2->setPackage('Net_FTP');
$p2->setSummary($summary);
$p2->setDescription($description);
$p2->setChannel('pear.php.net');

$p2->setPackageType('php');

$p2->addGlobalReplacement('package-info', '@package_version@', 'version');

$p2->generateContents();

$p2->setReleaseVersion($current_version);
$p2->setAPIVersion('1.4.0');
$p2->setReleaseStability($current_stability);
$p2->setAPIStability('stable');

$p2->setNotes($current_notes);

$p2->addGlobalReplacement('package-info', '@package_version@', 'version');

$p2->addInstallAs('tests/AllTests.php', 'AllTests.php');
$p2->addInstallAs('tests/Net_FTPTest.php', 'Net_FTPTest.php');
$p2->addInstallAs('tests/config.php.dist', 'config.php.dist');
$p2->addInstallAs('tests/testfile.dat', 'testfile.dat');
$p2->addInstallAs('tests/extensions.ini', 'extensions.ini');

$p2->addRelease();

$p2->addMaintainer('lead', 'jorrit', 'Jorrit Schippers', 'jschippers@php.net',
    'yes');
$p2->addMaintainer('lead', 'toby', 'Tobias Schlitt', 'toby@php.net', 'no');

$p2->setPhpDep('4.3.0');
$p2->setPearinstallerDep('1.3.0');

$p2->setLicense('PHP License', 'http://www.php.net/license');

$p2->addExtensionDep('required', 'ftp');

if (isset($_GET['make']) ||
    (isset($_SERVER['argv']) && @$_SERVER['argv'][1] == 'make')) {
    echo "Writing package file\n";
    $p2->writePackageFile();
} else {
    echo "Debugging package file\n";
    $p2->debugPackageFile();
}
?>
